Add Quantize and convertToFP16 support to large model loading

gcatron opened the pull request:
Summary:
This adds support for loading deferred weights when they need to be quantized or converted to FP16. 
This PR splits foldElemKindConversionIntoIO into Input and Output and modified Input pass to support selectively folding static placeholders only and updated the Provisioner to check if a conversion is needed before transferring the weight to the device.

Documentation:

Test Plan: added new FP16StaticPlaceholderInference test to test conversion flow.

jfix71 commented on lib/Runtime/Provisioner/Provisioner.cpp:
```diff
@@ -448,18 +448,32 @@ Error Provisioner::provision(DAGListTy &networks, Module &module,
+      // Convert the weight if needed.
+      auto newTy = PH->getType();
+      auto weight = loader->getTensor();
+      if (!weight->getType().isEqual(newTy)) {
```
This should only be the case if it's a static PH right? Maybe add an assert here?

gcatron commented:
@jfix71  Good point, I went ahead and renamed `onlyFoldStaticPlaceholderConversions` to `foldStaticPlaceholderConversions` and modified the fold inputs pass to skip non-static placeholders if `foldElemKindConversionIntoIO` is not set. Both options should work together now.

jfix71 commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -3374,11 +3399,22 @@ Error glow::optimizeFunction(Function *F, const Backend &B,
-                            {FunctionPassID::FoldElemKindConversionIntoIO});
+    FunctionPassManager FPM(
+        "FoldElemKindConversionIntoIO",
+        {{FunctionPassID::FoldElemKindConversionIntoInputs},
```
small nit: Looks a little funny/inefficient to me to back to back potentially create pipelines with `FoldElemKindConversionIntoInputs`, especially since they could both be run.

One thing you could do is merge this into a single `FunctionPassManager` -- check if `foldStaticPlaceholderConversions || foldElemKindConversionIntoIO`, create a `FunctionPassPipeline` with `FoldElemKindConversionIntoInputs` since it'll always be run, and then conditionally add `FoldElemKindConversionIntoOutputs` if `foldElemKindConversionIntoIO`. Then feed that `FunctionPassPipeline` into an `FPM`.

(whew that's a mouthful)
